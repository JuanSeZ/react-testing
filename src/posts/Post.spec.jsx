// Task 2: test the Post component.
// Acceptance criteria
// 1) Include at least one snapshot.
// 2) The copy to clipboard functionality must be fully tested.
import React from "react";
import { describe, expect, vi, test } from 'vitest';
import {usePost} from "@/posts/hooks/index.js";
import {fireEvent, getByText, render, waitFor} from "@testing-library/react";
import {Post} from "@/posts/Post.jsx";
import {MemoryRouter} from "react-router-dom";

vi.mock('./hooks/usePost', () => ({
    usePost: vi.fn(),
}));


vi.mock('lucide-react', async (importOriginal) => {
    const mod = await importOriginal()
    return {
        ...mod,
        Share2Icon: (props) => <div {...props}>Share2Icon</div>,
        ArrowBigLeftIcon: (props) => <div {...props}>ArrowBigLeftIcon</div>,
    }
})


describe('Post', () => {

    test('should match snapshot when there is succesful', () => {

        usePost.mockReturnValue({
            status: 'success',
            data: {
                title: 'Title',
                body: 'Body'
            }
        });

        const {container} = render(<Post/>, {wrapper: MemoryRouter});
        expect(container.firstChild).toMatchSnapshot();
    });

    test('should match snapshot when there is an error', () => {

        usePost.mockReturnValue({
            status: 'error',
            error: new Error('An error occurred!')
        });

        const {container} = render(<Post/>, {wrapper: MemoryRouter});
        expect(container.firstChild).toMatchSnapshot();
    });
})

describe('Post', () => {

    test('should copy to clipboard', async () => {
        window.navigator.clipboard = {writeText: vi.fn(() => Promise.resolve())}

        usePost.mockReturnValue({
            status: 'success',
            data: {
                title: '..',
                body: '..'
            }
        })

        const {container} = render(<Post />, { wrapper: (props) => <MemoryRouter initialEntries={['/posts/1']} {...props} /> })
        fireEvent.click(getByText(container,'Share2Icon'))

        await waitFor(() => {
            expect(navigator.clipboard.writeText).toHaveBeenCalledWith(`${location.origin}/posts/1`)
        })
    })

})