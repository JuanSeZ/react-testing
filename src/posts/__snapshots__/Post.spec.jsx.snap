// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Post > should match snapshot when there is an error 1`] = `
<div
  class="flex items-center"
>
  <a
    class="flex items-center"
    href="/posts"
  >
    <div>
      ArrowBigLeftIcon
    </div>
    <span>
      Return to posts
    </span>
  </a>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10 ml-12 cursor-pointer"
  >
    <div>
      Share2Icon
    </div>
  </button>
</div>
`;

exports[`Post > should match snapshot when there is succesful 1`] = `
<div
  class="flex items-center"
>
  <a
    class="flex items-center"
    href="/posts"
  >
    <div>
      ArrowBigLeftIcon
    </div>
    <span>
      Return to posts
    </span>
  </a>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10 ml-12 cursor-pointer"
  >
    <div>
      Share2Icon
    </div>
  </button>
</div>
`;
